### PR TITLE
Ensure all watchers are executed in scheduling order

### PR DESCRIPTION
**Pull request by philippjfr**
Fixes https://github.com/pyviz/param/issues/322

- [x] Add tests

### CODE REVIEW

**Comment by philippjfr** [comment]
One way to make this PR cleaner would be to add a context manager for setting and unsetting the BATCH_WATCH flag.

**Comment by philippjfr** [comment]
Ready to review/merge.

File: param/parameterized.py [hunk 1]
```diff
@@ -75,6 +75,22 @@ def logging_level(level):
         param_logger.setLevel(logging_level)
 
 
+@contextmanager
+def batch_watch(parameterized, run=True):
+    """
+    Context manager which batches the watchers on the parameterized.
+    If run=True it will dispatch the accumulated events on the watchers.
```
Comment (jbednar) [inline param/parameterized.py]: This docstring is a bit obscure; I can't tell what the purpose of this context manager is from it. Can you elaborate?

File: param/parameterized.py [hunk 2]
```diff
@@ -1215,6 +1238,7 @@ def set_param(self_, *args,**kwargs):
         positional arguments, but the keyword interface is preferred
         because it is more compact and can set multiple values.
         """
+        BATCH_WATCH = self_.self_or_cls.param._BATCH_WATCH
```
Comment (jbednar) [inline param/parameterized.py]: Why the scary capitals?

File: param/parameterized.py (hunk 1 above)
Comment (philippjfr) [reply param/parameterized.py]: Updated the docstring, hope it's clearer now.

File: param/parameterized.py (hunk 2 above)
Comment (philippjfr) [reply param/parameterized.py]: Just following the style of the original variable name, not sure it has to be capitalized but if we decide to change I would change it everywhere.

Comment (jbednar) [reply param/parameterized.py]: As you wish.

File: param/parameterized.py [hunk 3]
```diff
@@ -78,8 +78,11 @@ def logging_level(level):
 @contextmanager
 def batch_watch(parameterized, run=True):
     """
-    Context manager which batches the watchers on the parameterized.
-    If run=True it will dispatch the accumulated events on the watchers.
+    Context manager to batch watcher events on a parameterized object.
+    The context manager will queue any events triggered by setting a
+    parameter on the supplied parameterized object and dispatch them
+    all at once when the context manager exits. If run=False the
+    queued events are not dispatched and should be processed manually.
```
Comment (jbednar) [inline param/parameterized.py]: Much clearer, thanks!

**Comment by martindurant** [comment]
Quick question on this: does the watcher get the event itself in its arguments? I would say, that an event which is going to cause another like event of a different value to be spawned should explicitly set the event to be ignored downstream. That way, if you want a "less than 4" filter on an element as you described, other watchers *never* see the value 5, even temporarily. This leaves it up to the programmer to think about whether it makes sense for an event to propagate or not. 
(This is a separate concept from batching events as above)

**Comment by philippjfr** [comment]
>Quick question on this: does the watcher get the event itself in its arguments?

Yes it does, and I agree with you. I think a watcher should be able to invalidate an event it receives so that downstream watchers never see it. That said maybe a better solution would be to offer a way to define preprocessors on parameters so the value is validated before it is set and triggers downstream events. I can't quite envision if there are other scenarios where you might want a watcher to invalidate an event it receives.

**Comment by martindurant** [comment]
> to offer a way to define preprocessors on parameters so the value is validated before it is set

This way means having a different type of watcher, which may be confusing. I suppose it could be a very simple function valid(val) -> True/False.

I can imagine scenarios, such as the new value caused a procedure to fire which failed, so it is OK to set that value (it's not invalid, only failing at the current moment and set of other values) but follow-on procedures should not run. You could argue this case is moot, any follow-ons should respond the a change in a separate output area...

**Comment by philippjfr** [comment]
This PR also happens to fix https://github.com/pyviz/panel/issues/143.

I think we should merge this PR and then file an issue to discuss any suggestions related to letting users invalidate events.

**Comment by martindurant** [comment]
That's certainly fine with me.

**Comment by jlstevens** [comment]
I think I'll go ahead and merge this PR so we can test it. I agree with Jim that the earlier discussion suggests that this is the right approach.

**Pull request merged**